### ISSUE TITLE
Add `sched.feasibility` service in support of new job validation plugin

### ISSUE THREAD

**Issue by SteVwonder** [flux-framework/flux-sched#814]
https://github.com/flux-framework/flux-core/pull/3533 added a job validation plugin that pings a `sched.feasibility` service to determine if a job is feasible for scheduling at ingest time.

Support for this service was added to `sched-simple`, and it would be good to have similar functionality in Fluxion too.

**Comment by grondo** [comment]
Thanks, @SteVwonder!

Note that the `feasibility` validator plugin has an option to override the service name for the feasibility check, `--feasibility-service=NAME` in case Fluxion needs to send the RPC to the resource module (or another service) instead.

**Comment by dongahn** [comment]
FYI -- I will be looking into this as my next item.

**Comment by grondo** [comment]
> Can the Fluxion feasibility validator add a "feasibility validated" stamp to the jobspec?

Currently there is no way for the validator to modify, amend, or annotate the jobspec - validators are restricted to returning success (an `errnum` of `0`) or an error number and an error string.

Even in principle, Flux components cannot currently add anything to the jobspec since that would invalidate the user's signature on the jobspec.

Edit: Oh, after reading back your response, you may be talking some kind of external annotation of a job which Flux components could add. We have indeed talked about this before, but there is nothing currently that supports this.

I think what we'll want to do is have the feasibility plugin enabled by configuration and the scheduler could check for that same configuration to determine if a feasibility check has already been accomplished.

**Comment by dongahn** [comment]
> Even in principle, Flux components cannot currently add anything to the jobspec since that would invalidate the user's signature on the jobspec.

From rabbit support discussion, I somehow mistakenly thought this could be done. 

> Edit: Oh, after reading back your response, you may be talking some kind of external annotation of a job which Flux components could add. We have indeed talked about this before, but there is nothing currently that supports this.

From the scheduler point of view, if the fact that the job was already checked through the feasibility validator is somehow passed to `sched.alloc`, it can make use of it to avoid redundant feasibility...

> I think what we'll want to do is have the feasibility plugin enabled by configuration and the scheduler could check for that same configuration to determine if a feasibility check has already been accomplished.

Good point. This would work for the system instance. 

But my concern would be fluxion users might want to specialize the scheduling of the user level instance without the front-loaded feasibility since this can significantly lower their throughput...


Ok. It seems this would be the path to least resistance.

**Comment by dongahn** [comment]
> But my concern would be fluxion users might want to specialize the scheduling of the user level instance without the front-loaded feasibility since this can significantly lower their throughput...

Just to qualify, if the `match-allocate` succeed, we don't do feasibility so for "good input", schedule-loop time feasibility will outperform.

**Comment by grondo** [comment]
> From rabbit support discussion, I somehow mistakenly thought this could be done.

Yeah, we've discussed that it could be useful if job manager plugins and validator plugins could amend the jobspec. Specifically, the redacted jobspec held by the job manager could allow direct modification within reason. The job manager could then pass the jobspec directly to the scheduler in the `alloc` request, instead of having the scheduler separately fetch the jobspec out of the KVS. Thus the scheduler could get a hold of an amended jobspec, but the job shell would be using the original jobspec, so that the Flux instance could not directly tamper with the job commands, environment, etc.

However, this is not implemented yet. I wasn't sure where that requirement was on the roadmap.

**Comment by grondo** [comment]
Ah, I see the flag to bypass feasibility check is already in flux-framework/flux-core#3788. In this case a key added by the creator of the jobspec is suggested.

**Comment by grondo** [comment]
More generally, mutability of jobspec post submission is flux-framework/flux-core#3416

**Comment by dongahn** [comment]
Good to see these discussion threads. 

For my current short-term task, it seems just adding a fluxion feasibility validator seems to make sense. If either mutability of jobspec post submission or support for a  validator to annotate the job is supported, we can improve on it to further optimize scheduling performance.
